Rollbar Views for Project and Deploy

ragurney opened the pull request:
@zendesk/devex 
<img width="1224" alt="screen shot 2018-03-21 at 12 28 25 pm" src="https://user-images.githubusercontent.com/15261525/37733331-f32fc0c4-2d04-11e8-854b-8b10a54a7105.png">
<img width="1243" alt="screen shot 2018-03-21 at 12 28 05 pm" src="https://user-images.githubusercontent.com/15261525/37733332-f3311190-2d04-11e8-8fbd-45a29e255f33.png">
<img width="1204" alt="screen shot 2018-03-21 at 12 42 23 pm" src="https://user-images.githubusercontent.com/15261525/37733460-56be6bd6-2d05-11e8-900b-71afab78ca00.png">
<img width="1177" alt="screen shot 2018-03-21 at 12 41 40 pm" src="https://user-images.githubusercontent.com/15261525/37733462-57fc0d1e-2d05-11e8-9bf5-a1b96370b06b.png">
<img width="1210" alt="screen shot 2018-03-21 at 12 40 56 pm" src="https://user-images.githubusercontent.com/15261525/37733463-591c6acc-2d05-11e8-92f1-304d532ce69c.png">

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb:
```diff
@@ -0,0 +1,29 @@
+<%
+  occurrence_term = 'occurrences_count'
```
makes more sense inside the if

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb:
```diff
@@ -0,0 +1,29 @@
+  if deploy.succeeded?
+%>
+  <% if deploy.project.rollbar_dashboards_settings.empty? %>
+    <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: deploy.project } %>
```
only projects that opt-in should get any UI change

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb:
```diff
@@ -0,0 +1,29 @@
+    <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: deploy.project } %>
+  <%
+    else
+      @data = deploy.project.rollbar_dashboards_settings.map do |dashboard|
```
maybe class name `RollbarDashboard` makes more sense ?

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb:
```diff
@@ -0,0 +1,29 @@
+        }
+
+        items = RollbarDashboards::Client.new(dashboard).items(params)
+        [dashboard, items&.sort_by { |i| i[occurrence_term] }&.reverse&.first(SamsonRollbarDashboards::Engine::NUM_ITEMS)]
```
prefer .fetch

grosser replied:
can sort by `-` to avoid the reverse ... or use `.last`

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb:
```diff
@@ -0,0 +1,29 @@
+    else
+      @data = deploy.project.rollbar_dashboards_settings.map do |dashboard|
+        params = {
+          date_from: deploy.created_at.in_time_zone(dashboard.time_zone).strftime('%m-%d-%Y %H:%M:%S'),
```
would prefer if only users have a time zone and dashboards have not

grosser replied:
... we got time display helpers for that flying around somewhere

ragurney replied:
The time zone of the Rollbar project is independent of the time zone of the user, so we'd want to store that separately in order to get the correct items for the deploy dashboard.

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb:
```diff
@@ -0,0 +1,29 @@
+        collection: @data,
+        as: :dashboard_data,
+        locals: {
+          title: "Top #{SamsonRollbarDashboards::Engine::NUM_ITEMS} Items That Occurred This Deploy",
```
items -> errors ?

"Top 5 Errors During This Deploy"

ragurney replied:
Rollbar uses the term 'Item' in it's dashboard. Do we want to copy their terminology or would 'Error' be more intuitive?

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_project.html.erb:
```diff
@@ -0,0 +1,21 @@
+<% if project.rollbar_dashboards_settings.empty? %>
+  <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: project } %>
```
same here, no UI change if users did not opt in

grosser commented on plugins/rollbar_dashboards/app/views/samson_rollbar_dashboards/_fields.html.erb:
```diff
@@ -0,0 +1,37 @@
+  <legend>
+    Rollbar Dashboards
+  </legend>
+  <% project = form.object %>
```
`@project` is also available

grosser commented on plugins/rollbar_dashboards/config/routes.rb:
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
```
where is that used ?

ragurney replied:
Remanent of a previous version of this feature 😅

grosser commented on plugins/rollbar_dashboards/app/models/rollbar_dashboards/setting.rb:
```diff
@@ -0,0 +1,19 @@
+
+    validates :base_url, :read_token, :time_zone, presence: true
+
+    def resolved_read_token
```
if this is used on every project show/deploy-show then we might want to cache that ... also not blow up and make the project undeployable in cas vault has a hickup

ragurney replied:
If someone were to update the token and the token was cached then the feature would be broken until the cached value expired? I could see that leading to some confusion. I could cache the result to the `RollbarDashboards::Client` call on the deploy page as well? Then the entire dashboard is essentially cached for 2 minutes, including the token.

grosser commented on plugins/rollbar_dashboards/lib/rollbar_dashboards/client.rb:
```diff
@@ -0,0 +1,40 @@
+        data = ::JSON.parse(response.body)
+        data['result'].map { |item_hash| item_hash['item'] }
+      end
+    rescue ::JSON::ParserError
```
need to rescue more here ... send to ErrorNotifier and return nil ... we can't blow up peoples project/deploy pages
... and needs a timeout ideally at the http level since `Timeout.timeout` is dangerous

grosser commented on plugins/rollbar_dashboards/lib/rollbar_dashboards/client.rb:
```diff
@@ -0,0 +1,40 @@
+    end
+
+    def items(additional_params = {})
+      params = {
```
since additional_params is not reused I would normally do something like:

```
params = {foo: bar}.merge(params)
````

grosser replied:
also don't name different data-structures the same .. params -> hash ... query -> string

grosser commented on plugins/rollbar_dashboards/lib/rollbar_dashboards/client.rb:
```diff
@@ -0,0 +1,40 @@
+
+      response = Faraday.get("#{@base}/items/?#{params}")
+      if response.success?
+        data = ::JSON.parse(response.body)
```
prefer parsing json with `symbolize_names: true` so everything is symbols ✨

grosser commented on plugins/rollbar_dashboards/lib/samson_rollbar_dashboards/samson_plugin.rb:
```diff
@@ -0,0 +1,28 @@
+  class Engine < Rails::Engine
+    NUM_ITEMS = 4
+
+    initializer "refinery.assets.precompile" do |app|
```
refinery ?

dragonfax commented:
All-in-all. Very cool.

On Wed, Mar 21, 2018 at 1:16 PM Michael Grosser <notifications@github.com>
wrote:

> *@grosser* requested changes on this pull request.
>
> looks pretty good!
> ------------------------------
>
> In plugins/rollbar_dashboards/README.md
> <https://github.com/zendesk/samson/pull/2660#discussion_r176217927>:
>
> > @@ -0,0 +1,3 @@
> +# Rollbar Plugin
>
> Rollbar Dashboard Plugin
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/helpers/rollbar_dashboards/application_helper.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176218524>:
>
> > @@ -0,0 +1,15 @@
> +# frozen_string_literal: true
> +
> +module RollbarDashboards
> +  module ApplicationHelper
> +    def badge_color_class(occurrences)
>
> this is very different for each project ... maybe simplify by not showing
> any colors ?
> ------------------------------
>
> In plugins/rollbar_dashboards/app/models/rollbar_dashboards/setting.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176218667>:
>
> > @@ -0,0 +1,19 @@
> +# frozen_string_literal: true
> +
> +module RollbarDashboards
> +  class Setting < ActiveRecord::Base
> +    self.table_name = 'rollbar_dashboards_settings'
> +
> +    belongs_to :project
> +
> +    validates :base_url, :read_token, :time_zone, presence: true
> +
> +    def resolved_read_token
> +      if key = read_token.to_s.dup.sub!(TerminalExecutor::SECRET_PREFIX, "")
>
> don't copy-paste ... extract method :D
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/rollbar_dashboards/_project.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176219059>:
>
> > @@ -0,0 +1,21 @@
> +<% if project.rollbar_dashboards_settings.empty? %>
> +  <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: project } %>
> +<%
> +  else
> +    @data = project.rollbar_dashboards_settings.map do |dashboard|
>
> instance vars are shared, so make the name specific to avoid collisions
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176219313>:
>
> > @@ -0,0 +1,29 @@
> +<%
> +  occurrence_term = 'occurrences_count'
>
> makes more sense inside the if
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176219527>:
>
> > @@ -0,0 +1,29 @@
> +<%
> +  occurrence_term = 'occurrences_count'
> +  if deploy.succeeded?
> +%>
> +  <% if deploy.project.rollbar_dashboards_settings.empty? %>
> +    <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: deploy.project } %>
>
> only projects that opt-in should get any UI change
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176219733>:
>
> > @@ -0,0 +1,29 @@
> +<%
> +  occurrence_term = 'occurrences_count'
> +  if deploy.succeeded?
> +%>
> +  <% if deploy.project.rollbar_dashboards_settings.empty? %>
> +    <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: deploy.project } %>
> +  <%
> +    else
> +      @data = deploy.project.rollbar_dashboards_settings.map do |dashboard|
>
> maybe class name RollbarDashboard makes more sense ?
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176219867>:
>
> > +<%
> +  occurrence_term = 'occurrences_count'
> +  if deploy.succeeded?
> +%>
> +  <% if deploy.project.rollbar_dashboards_settings.empty? %>
> +    <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: deploy.project } %>
> +  <%
> +    else
> +      @data = deploy.project.rollbar_dashboards_settings.map do |dashboard|
> +        params = {
> +          date_from: deploy.created_at.in_time_zone(dashboard.time_zone).strftime('%m-%d-%Y %H:%M:%S'),
> +          date_to: deploy.next_successful_deploy&.created_at&.in_time_zone(dashboard.time_zone)&.strftime('%m-%d-%Y %H:%M:%S') || '',
> +        }
> +
> +        items = RollbarDashboards::Client.new(dashboard).items(params)
> +        [dashboard, items&.sort_by { |i| i[occurrence_term] }&.reverse&.first(SamsonRollbarDashboards::Engine::NUM_ITEMS)]
>
> prefer .fetch
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176220066>:
>
> > @@ -0,0 +1,29 @@
> +<%
> +  occurrence_term = 'occurrences_count'
> +  if deploy.succeeded?
> +%>
> +  <% if deploy.project.rollbar_dashboards_settings.empty? %>
> +    <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: deploy.project } %>
> +  <%
> +    else
> +      @data = deploy.project.rollbar_dashboards_settings.map do |dashboard|
> +        params = {
> +          date_from: deploy.created_at.in_time_zone(dashboard.time_zone).strftime('%m-%d-%Y %H:%M:%S'),
>
> would prefer if only users have a time zone and dashboards have not
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176220165>:
>
> > @@ -0,0 +1,29 @@
> +<%
> +  occurrence_term = 'occurrences_count'
> +  if deploy.succeeded?
> +%>
> +  <% if deploy.project.rollbar_dashboards_settings.empty? %>
> +    <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: deploy.project } %>
> +  <%
> +    else
> +      @data = deploy.project.rollbar_dashboards_settings.map do |dashboard|
> +        params = {
> +          date_from: deploy.created_at.in_time_zone(dashboard.time_zone).strftime('%m-%d-%Y %H:%M:%S'),
>
> ... we got time display helpers for that flying around somewhere
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176220365>:
>
> > +<%
> +  occurrence_term = 'occurrences_count'
> +  if deploy.succeeded?
> +%>
> +  <% if deploy.project.rollbar_dashboards_settings.empty? %>
> +    <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: deploy.project } %>
> +  <%
> +    else
> +      @data = deploy.project.rollbar_dashboards_settings.map do |dashboard|
> +        params = {
> +          date_from: deploy.created_at.in_time_zone(dashboard.time_zone).strftime('%m-%d-%Y %H:%M:%S'),
> +          date_to: deploy.next_successful_deploy&.created_at&.in_time_zone(dashboard.time_zone)&.strftime('%m-%d-%Y %H:%M:%S') || '',
> +        }
> +
> +        items = RollbarDashboards::Client.new(dashboard).items(params)
> +        [dashboard, items&.sort_by { |i| i[occurrence_term] }&.reverse&.first(SamsonRollbarDashboards::Engine::NUM_ITEMS)]
>
> can sort by - to avoid the reverse ... or use .last
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176220585>:
>
> > +      @data = deploy.project.rollbar_dashboards_settings.map do |dashboard|
> +        params = {
> +          date_from: deploy.created_at.in_time_zone(dashboard.time_zone).strftime('%m-%d-%Y %H:%M:%S'),
> +          date_to: deploy.next_successful_deploy&.created_at&.in_time_zone(dashboard.time_zone)&.strftime('%m-%d-%Y %H:%M:%S') || '',
> +        }
> +
> +        items = RollbarDashboards::Client.new(dashboard).items(params)
> +        [dashboard, items&.sort_by { |i| i[occurrence_term] }&.reverse&.first(SamsonRollbarDashboards::Engine::NUM_ITEMS)]
> +      end
> +  %>
> +    <%=
> +      render partial: 'rollbar_dashboards/dashboard',
> +        collection: @data,
> +        as: :dashboard_data,
> +        locals: {
> +          title: "Top #{SamsonRollbarDashboards::Engine::NUM_ITEMS} Items That Occurred This Deploy",
>
> items -> errors ?
>
> "Top 5 Errors During This Deploy"
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/rollbar_dashboards/_project.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176220885>:
>
> > @@ -0,0 +1,21 @@
> +<% if project.rollbar_dashboards_settings.empty? %>
> +  <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: project } %>
>
> same here, no UI change if users did not opt in
> ------------------------------
>
> In
> plugins/rollbar_dashboards/app/views/samson_rollbar_dashboards/_fields.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176221229>:
>
> > @@ -0,0 +1,37 @@
> +<fieldset>
> +  <legend>
> +    Rollbar Dashboards
> +  </legend>
> +  <% project = form.object %>
>
> @project is also available
> ------------------------------
>
> In plugins/rollbar_dashboards/config/routes.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176221650>:
>
> > @@ -0,0 +1,12 @@
> +# frozen_string_literal: true
>
> where is that used ?
> ------------------------------
>
> In plugins/rollbar_dashboards/app/models/rollbar_dashboards/setting.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176222202>:
>
> > @@ -0,0 +1,19 @@
> +# frozen_string_literal: true
> +
> +module RollbarDashboards
> +  class Setting < ActiveRecord::Base
> +    self.table_name = 'rollbar_dashboards_settings'
> +
> +    belongs_to :project
> +
> +    validates :base_url, :read_token, :time_zone, presence: true
> +
> +    def resolved_read_token
>
> if this is used on every project show/deploy-show then we might want to
> cache that ... also not blow up and make the project undeployable in cas
> vault has a hickup
> ------------------------------
>
> In plugins/rollbar_dashboards/lib/rollbar_dashboards/client.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176222534>:
>
> > +      @token = dashboard.resolved_read_token
> +    end
> +
> +    def top_errors(hours: 24, environments: ['production'])
> +      args = {
> +        access_token: @token,
> +        hours: hours,
> +        environments: environments.join(',')
> +      }.to_query
> +
> +      response = Faraday.get("#{@base}/reports/top_active_items?#{args}")
> +      if response.success?
> +        data = ::JSON.parse(response.body)
> +        data['result'].map { |item_hash| item_hash['item'] }
> +      end
> +    rescue ::JSON::ParserError
>
> need to rescue more here ... send to ErrorNotifier and return nil ... we
> can't blow up peoples project/deploy pages
> ... and needs a timeout ideally at the http level since Timeout.timeout
> is dangerous
> ------------------------------
>
> In plugins/rollbar_dashboards/lib/rollbar_dashboards/client.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176222776>:
>
> > +        access_token: @token,
> +        hours: hours,
> +        environments: environments.join(',')
> +      }.to_query
> +
> +      response = Faraday.get("#{@base}/reports/top_active_items?#{args}")
> +      if response.success?
> +        data = ::JSON.parse(response.body)
> +        data['result'].map { |item_hash| item_hash['item'] }
> +      end
> +    rescue ::JSON::ParserError
> +      nil
> +    end
> +
> +    def items(additional_params = {})
> +      params = {
>
> since additional_params is not reused I would normally do something like:
>
> params = {foo: bar}.merge(params)
>
> ------------------------------
>
> In plugins/rollbar_dashboards/lib/rollbar_dashboards/client.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176222935>:
>
> > +        access_token: @token,
> +        hours: hours,
> +        environments: environments.join(',')
> +      }.to_query
> +
> +      response = Faraday.get("#{@base}/reports/top_active_items?#{args}")
> +      if response.success?
> +        data = ::JSON.parse(response.body)
> +        data['result'].map { |item_hash| item_hash['item'] }
> +      end
> +    rescue ::JSON::ParserError
> +      nil
> +    end
> +
> +    def items(additional_params = {})
> +      params = {
>
> also don't name different data-structures the same .. params -> hash ...
> query -> string
> ------------------------------
>
> In plugins/rollbar_dashboards/lib/rollbar_dashboards/client.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176223087>:
>
> > +      if response.success?
> +        data = ::JSON.parse(response.body)
> +        data['result'].map { |item_hash| item_hash['item'] }
> +      end
> +    rescue ::JSON::ParserError
> +      nil
> +    end
> +
> +    def items(additional_params = {})
> +      params = {
> +        access_token: @token
> +      }.merge(additional_params).to_query
> +
> +      response = Faraday.get("#{@base}/items/?#{params}")
> +      if response.success?
> +        data = ::JSON.parse(response.body)
>
> prefer parsing json with symbolize_names: true so everything is symbols ✨
> ------------------------------
>
> In
> plugins/rollbar_dashboards/lib/samson_rollbar_dashboards/samson_plugin.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176223227>:
>
> > @@ -0,0 +1,28 @@
> +# frozen_string_literal: true
> +
> +module SamsonRollbarDashboards
> +  class Engine < Rails::Engine
> +    NUM_ITEMS = 4
> +
> +    initializer "refinery.assets.precompile" do |app|
>
> refinery ?
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/2660#pullrequestreview-105883192>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4_rhCO5igJT4o42oZs5J6Maek_dRlwks5tgrT4gaJpZM4S1hdz>
> .
>

grosser commented:
if the dates returned from rollbar are in utc then we don't need to worry
about that ?

On Wed, Mar 21, 2018 at 1:38 PM, Ryan Gurney <notifications@github.com>
wrote:

> *@ragurney* commented on this pull request.
> ------------------------------
>
> In plugins/rollbar_dashboards/app/views/rollbar_dashboards/_
> deploy.html.erb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176229905>:
>
> > @@ -0,0 +1,29 @@
> +<%
> +  occurrence_term = 'occurrences_count'
> +  if deploy.succeeded?
> +%>
> +  <% if deploy.project.rollbar_dashboards_settings.empty? %>
> +    <%= render partial: 'rollbar_dashboards/no_dashboards', locals: { project: deploy.project } %>
> +  <%
> +    else
> +      @data = deploy.project.rollbar_dashboards_settings.map do |dashboard|
> +        params = {
> +          date_from: deploy.created_at.in_time_zone(dashboard.time_zone).strftime('%m-%d-%Y %H:%M:%S'),
>
> The time zone of the Rollbar project is independent of the time zone of
> the user, so we'd want to store that separately in order to get the correct
> items for the deploy dashboard.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/2660#discussion_r176229905>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZzOQiSkml1Ov3anxkvkMwisKU5SYks5tgro3gaJpZM4S1hdz>
> .
>

grosser commented:
ah yeah then that should be enough caching ...

On Wed, Mar 21, 2018 at 2:22 PM, Ryan Gurney <notifications@github.com>
wrote:

> *@ragurney* commented on this pull request.
> ------------------------------
>
> In plugins/rollbar_dashboards/app/models/rollbar_dashboards/setting.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176242467>:
>
> > @@ -0,0 +1,19 @@
> +# frozen_string_literal: true
> +
> +module RollbarDashboards
> +  class Setting < ActiveRecord::Base
> +    self.table_name = 'rollbar_dashboards_settings'
> +
> +    belongs_to :project
> +
> +    validates :base_url, :read_token, :time_zone, presence: true
> +
> +    def resolved_read_token
>
> If someone were to update the token and the token was cached then the
> feature would be broken until the cached value expired? I could see that
> leading to some confusion. I could cache the result to the
> RollbarDashboards::Client call on the deploy page as well? Then the
> entire dashboard is essentially cached for 2 minutes, including the token.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/2660#discussion_r176242467>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ00S2OM7TAOVEp6TQhMBxrwixkuXks5tgsR_gaJpZM4S1hdz>
> .
>

ragurney commented:
The dates in Rollbar aren't in UTC. From my understanding the query params are used as if they are in whatever time zone the project is set to 😕

grosser commented:
then play with it a bit to make sure ... or check api docs ... ideally we'd
automatically fill the timezone value in ...
but then still: convert to utc and display with out time helpers

On Wed, Mar 21, 2018 at 2:50 PM, Ryan Gurney <notifications@github.com>
wrote:

> The dates in Rollbar aren't in UTC. From my understanding the query params
> are used as if they are in whatever time zone the project is set to 😕
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/2660#issuecomment-375107640>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZxZJeogfRBfyw0QD4EIck7t2IROfks5tgssagaJpZM4S1hdz>
> .
>

ragurney commented:
Addressed comments, but I'm going to have to re-examine the logic of getting the items by date range. I'm getting some odd results that weren't expected. Might have to resort to RQL to get it to work.

ragurney commented:
After some more digging, confirmed I'll have to use RQL (create a job, then get the job result in two separate requests). The method I was using previously isn't officially supported and gives erroneous results.

dragonfax commented:
weird.

On Wed, Mar 21, 2018 at 5:25 PM Ryan Gurney <notifications@github.com>
wrote:

> After some more digging, confirmed I'll have to use RQL (create a job,
> then get the job result in two separate requests). The method I was using
> previously isn't officially supported and gives erroneous results.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/2660#issuecomment-375138992>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA4_rpLpfRXg3d_3dywLP_J5WN_Xhgvkks5tgu-UgaJpZM4S1hdz>
> .
>

grosser commented on lib/samson/model_utility_methods.rb:
```diff
@@ -0,0 +1,14 @@
+
+module Samson
+  module ModelUtilityMethods
+    def self.resolved_attribute(model_instance, attribute, project)
```
pass in attribute_value instead of instance+ attribute

grosser replied:
can live on the KeyResolver

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb:
```diff
@@ -0,0 +1,23 @@
+<% unless deploy.project.rollbar_dashboards_settings.empty? || !deploy.succeeded?
```
prefer to use if instead of unless with multiple conditions

`if settings.any? && deploy.succeeded?`

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb:
```diff
@@ -0,0 +1,23 @@
+<% unless deploy.project.rollbar_dashboards_settings.empty? || !deploy.succeeded?
+    occurrence_term = :occurrences_count
+    @data = deploy.project.rollbar_dashboards_settings.map do |dashboard|
+      params = {
```
maybe the dashboard could know how to get items ?

`dashboard.items(deploy)`

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_no_dashboards.html.erb:
```diff
@@ -0,0 +1,5 @@
+<div class="alert alert-info alert-dismissible" role="alert">
```
no longer used

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_project.html.erb:
```diff
@@ -0,0 +1,19 @@
+<%
+  unless project.rollbar_dashboards_settings.empty?
```
avoid double negative -> `if any`

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_project.html.erb:
```diff
@@ -0,0 +1,19 @@
+<%
+  unless project.rollbar_dashboards_settings.empty?
+    @dashboard_data = project.rollbar_dashboards_settings.map do |dashboard|
+      items = Rails.cache.fetch("project_#{dashboard.cache_key}", expires_in: 10.minutes) do
```
FYI passing models to `.fetch` uses cache_key on them, so can do `.fetch(['project', deployment], expires_in : 10.minutes)` ...

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_project.html.erb:
```diff
@@ -0,0 +1,19 @@
+      end
+      [dashboard, items]
+    end
+%>
```
might be simpler to cache the html instead of caching the data and then passing it in
then it could fetch the data in the partial

grosser commented:
A deploy should only show items from the environment it deploys to.
So dashboard should be on stage ... and project should display top-errors from the first stage sorted by `production?` right ?

ragurney commented:
I think it makes sense to have the dashboard for a deploy live on the deploy show page? That way you could go back and look at a past deploy and its errors, which is how the desired functionality/design described to me. I'm not sure what you mean by: 
> and project should display top-errors from the first stage sorted by production

Right now, the project dashboard is displaying the top errors in the production environment for it's rollbar project which is what we'd want, right? Just a summary of the current state of production

ragurney commented:
Deploys should display items only for it's environment. I think Jason and I talked about that being a later PR, but I could incorporate it into this RQL stuff.

grosser commented:
the deploy does not know it's environment (master/staging, production works ... if users use strange names then that fails ...) so we could configure it on the stage, that's how the rollbar_hook plugin works ... having to copy-paste the config to multiple stages would be cumbersome though :/

telling master deployers that they caused production errors is misleading ... so either don't show unless it's production or say "errors in all environments" ... maybe we can get environment from the deploy groups `Environment` it is deploying too ...

grosser commented:
I guess since it's opt-in it does not really matter and we can play around with it ...

dragonfax commented:
The project page can show just production, (or everything). Either way is
fine with me. But I think production is more relavant for the project as a
whole.

The stage page, should show errors from the environment that it deploys to.
You can find this by following Stage->deploy_groups->first->environment



On Wed, Mar 21, 2018 at 6:21 PM Ryan Gurney <notifications@github.com>
wrote:

> Deploys should display items only for it's environment. I think Jason and
> I talked about that being a later PR, but I could incorporate it into this
> RQL stuff.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/2660#issuecomment-375147482>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA4_rhybuldeQw3xu4fXnG6W0_2zD3Yqks5tgvyEgaJpZM4S1hdz>
> .
>

dragonfax commented:
deploy groups know which environment they are in.

On Wed, Mar 21, 2018 at 6:29 PM Jason Stillwell <jstillwell@zendesk.com>
wrote:

> The project page can show just production, (or everything). Either way is
> fine with me. But I think production is more relavant for the project as a
> whole.
>
> The stage page, should show errors from the environment that it deploys
> to. You can find this by following Stage->deploy_groups->first->environment
>
>
>
> On Wed, Mar 21, 2018 at 6:21 PM Ryan Gurney <notifications@github.com>
> wrote:
>
>> Deploys should display items only for it's environment. I think Jason and
>> I talked about that being a later PR, but I could incorporate it into this
>> RQL stuff.
>>
>> —
>> You are receiving this because you are on a team that was mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/zendesk/samson/pull/2660#issuecomment-375147482>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AA4_rhybuldeQw3xu4fXnG6W0_2zD3Yqks5tgvyEgaJpZM4S1hdz>
>> .
>>
>

ragurney commented:
Updated deploy dashboard logic, now with RQL and async job result loading ✨

grosser commented on plugins/rollbar_dashboards/app/controllers/rollbar_dashboards/dashboards_controller.rb:
```diff
@@ -0,0 +1,64 @@
+  def wait_for_rql_job_results(job_id, client)
+    MAX_RQL_JOB_WAIT_TIME.times do
+      result = client.rql_job_result(job_id)
+      return result if result.nil? || result != RollbarDashboards::Client::RQL_JOB_WAITING_TERM
```
rql_job_result should return nil when the result is RQL_JOB_WAITING_TERM ? <-> keep logic local

ragurney replied:
So move the waiting logic into `rql_job_result`?

grosser commented on plugins/rollbar_dashboards/app/controllers/rollbar_dashboards/dashboards_controller.rb:
```diff
@@ -0,0 +1,64 @@
+    MAX_RQL_JOB_WAIT_TIME.times do
+      result = client.rql_job_result(job_id)
+      return result if result.nil? || result != RollbarDashboards::Client::RQL_JOB_WAITING_TERM
+      sleep(1)
```
it's kinda strange to sleep and then stop the loop :D 
`times do |i| ... break if i >= (MAX_RQL_JOB_WAIT_TIME - 1)`

grosser commented on plugins/rollbar_dashboards/app/controllers/rollbar_dashboards/dashboards_controller.rb:
```diff
@@ -0,0 +1,64 @@
+             AND #{timestamp_query}
+      GROUP  BY 1,
+                item.counter
+      ORDER  BY 4 DESC
```
can we put the limit here ?

grosser commented on plugins/rollbar_dashboards/app/models/rollbar_dashboards/setting.rb:
```diff
@@ -0,0 +1,13 @@
+
+module RollbarDashboards
+  class Setting < ActiveRecord::Base
+    NUM_ITEMS = 4
```
the model should not have display logic ... maybe better in the controller ?

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_dashboard.html.erb:
```diff
@@ -0,0 +1,21 @@
+    <%= "#{title} (#{dashboard_data.first.base_url})" %>
+  </div>
+  <% items = dashboard_data.second %>
+  <% if items %>
```
FYI could turn this into a single `if items = ` line

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_dashboard.html.erb:
```diff
@@ -0,0 +1,21 @@
+    </table>
+  <% else %>
+    <div class="panel-body">
+      <p>There are no items to display at this time...</p>
```
`to display at this time` -> `found` / `occurred`

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_deploy.html.erb:
```diff
@@ -0,0 +1,3 @@
+<% if deploy.project.rollbar_dashboards_settings.any? && deploy.succeeded? %>
```
a comment here like `<%# will be replaced by foobar.js %>` could be nice

grosser commented on plugins/rollbar_dashboards/app/views/rollbar_dashboards/_project.html.erb:
```diff
@@ -0,0 +1,18 @@
+  if project.rollbar_dashboards_settings.any?
+    @dashboard_data = project.rollbar_dashboards_settings.map do |dashboard|
+      items = Rails.cache.fetch(['project', dashboard], expires_in: 10.minutes) do
+        RollbarDashboards::Client.new(dashboard).top_errors&.first(RollbarDashboards::Setting::NUM_ITEMS)
```
this could potentially cache nil ... which means the cache won't work ...
might be best to move this somewhere in the helper/model/client code to properly test it too

grosser replied:
possibly also good to make this a lazy-load so the page stays fast and everything has the same workflow

grosser commented on plugins/rollbar_dashboards/app/views/samson_rollbar_dashboards/_fields.html.erb:
```diff
@@ -0,0 +1,29 @@
+
+  <div class="col-lg-offset-2">
+    <%= form.fields_for :rollbar_dashboards_settings do |setting| %>
+
```
extra newline

grosser commented on plugins/rollbar_dashboards/app/views/samson_rollbar_dashboards/_fields.html.erb:
```diff
@@ -0,0 +1,29 @@
+        <div class="col-lg-3">
+          <%= setting.text_field :read_token, class: "form-control", placeholder: "Project Read Token" %>
+        </div>
+        <div class="col-lg-10" style="margin-left: .5em">
```
size 10 for a delete checkbox seems strange

grosser commented:
yeah

On Fri, Mar 23, 2018 at 1:24 PM, Ryan Gurney <notifications@github.com>
wrote:

> *@ragurney* commented on this pull request.
> ------------------------------
>
> In plugins/rollbar_dashboards/app/controllers/rollbar_
> dashboards/dashboards_controller.rb
> <https://github.com/zendesk/samson/pull/2660#discussion_r176854159>:
>
> > +    end
> +
> +    render(
> +      partial: 'rollbar_dashboards/dashboard',
> +      collection: data,
> +      as: :dashboard_data,
> +      locals: { title: "Top #{RollbarDashboards::Setting::NUM_ITEMS} Items That Occurred This Deploy" }
> +    )
> +  end
> +
> +  private
> +
> +  def wait_for_rql_job_results(job_id, client)
> +    MAX_RQL_JOB_WAIT_TIME.times do
> +      result = client.rql_job_result(job_id)
> +      return result if result.nil? || result != RollbarDashboards::Client::RQL_JOB_WAITING_TERM
>
> So move the waiting logic into rql_job_result?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/2660#discussion_r176854159>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZz02SjkIp_w8bODvX-cKw5FQfYFMks5thVoDgaJpZM4S1hdz>
> .
>

grosser commented on app/assets/javascripts/responsive_load.js.erb:
```diff
@@ -1,14 +1,26 @@
 // show loading animation and errors inside the container that is being replaced
+$(document).ready(function () {
+  $('.lazy-load').each(function () {
```
`$('.lazy-load'). responsiveLoad()` is not the same ?

grosser commented on lib/samson/form_builder.rb:
```diff
@@ -54,6 +54,24 @@ def actions(delete: false, label: 'Save', &block)
     end
 
+    # Creates multi-row input field
+    def fields_for_many(association, description, add_rows_allowed: false)
```
good stuff! ❤️

grosser commented on plugins/flowdock/app/views/samson_flowdock/_fields.html.erb:
```diff
@@ -1,21 +1,22 @@
-  <% stage.flowdock_flows.build if stage.flowdock_flows.last.try(:name).blank? %>
+  <% 
+    stage = form.object
+    stage.flowdock_flows.build if stage.flowdock_flows.last.try(:name).blank? 
```
`unless stage.flowdock_flows.last&.new_record?` maybe ?

grosser commented on plugins/rollbar_dashboards/app/controllers/rollbar_dashboards/dashboards_controller.rb:
```diff
@@ -0,0 +1,76 @@
+  NUM_ITEMS = 4
+
+  def project_dashboard
+    project = Project.find_by_permalink(params.require(:project_id))
```
`find_by_permalink!` since we don't want nils

grosser commented on plugins/rollbar_dashboards/app/controllers/rollbar_dashboards/dashboards_controller.rb:
```diff
@@ -0,0 +1,76 @@
+  def deploy_dashboard
+    deploy = Deploy.find(params.require(:deploy_id))
+
+    dashboard_html = Rails.cache.fetch(deploy, expires_in: 10.minutes) do
```
this needs to be a more unique cache-key like `["deploy_dashboard", deploy]`

grosser commented on plugins/rollbar_dashboards/app/controllers/rollbar_dashboards/dashboards_controller.rb:
```diff
@@ -0,0 +1,76 @@
+  def project_dashboard
+    project = Project.find_by_permalink(params.require(:project_id))
+
+    dashboard_html = Rails.cache.fetch(project, expires_in: 10.minutes) do
```
needs a more unique cache key

grosser commented on plugins/rollbar_dashboards/lib/rollbar_dashboards/client.rb:
```diff
@@ -0,0 +1,84 @@
+
+      if response.success?
+        data = ::JSON.parse(response.body, symbolize_names: true)
+        data[:result].map { |item_hash| item_hash[:item] }
```
FYI usually nice to avoid nils via `data.fetch(:result).map { |i| i.fetch(:item) }`

grosser commented on plugins/rollbar_dashboards/lib/rollbar_dashboards/client.rb:
```diff
@@ -0,0 +1,84 @@
+      }.to_query
+
+      response = Faraday.get("#{@base}/reports/top_active_items?#{args}") do |req|
+        req.options.timeout = req.options.open_timeout = 1
```
might want to make that a `ROLLBAR_API_TIMEOUT` constant
